[claude-opus-4.7] The `ubuntu-24.04 Java 21` shard of the [system-tests run](https://github.com/apache/nifi/actions/runs/25525725112/job/74920788653) failed on the known flake `ContentClaimTruncationIT.testReplayedFlowFileContentNotTruncated`:

```
[ERROR] org.apache.nifi.tests.system.repositories.ContentClaimTruncationIT.testReplayedFlowFileContentNotTruncated -- Time elapsed: 13.58 s <<< ERROR!
org.apache.nifi.toolkit.client.NiFiClientException: Error retrieving FlowFile Content: Unable to find the specified content.
	at org.apache.nifi.tests.system.NiFiClientUtil.getFlowFileContent(NiFiClientUtil.java:2023)
	at org.apache.nifi.tests.system.NiFiClientUtil.getFlowFileContentAsByteArray(NiFiClientUtil.java:2001)
	at org.apache.nifi.tests.system.repositories.ContentClaimTruncationIT.testReplayedFlowFileContentNotTruncated(ContentClaimTruncationIT.java:346)
Caused by: jakarta.ws.rs.NotFoundException: HTTP 404 Not Found
```

Tracked under [NIFI-15903](https://github.com/apache/nifi/pull/11203) ("Flaky tests ClusteredConnectorIT and ContentClaimTruncationIT"). The failure is unrelated to this PR, which only changes `StandardStateMap` in `nifi-framework-components`. I have rerun the failed shard.

--- a/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/state/StandardStateMap.java
+++ b/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/state/StandardStateMap.java
@@ -20,6 +20,7 @@ package org.apache.nifi.controller.state;
 import org.apache.nifi.components.state.StateMap;
 
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
 
@@ -30,7 +31,12 @@ public class StandardStateMap implements StateMap {
     private final Optional<String> stateVersion;
 
     public StandardStateMap(final Map<String, String> stateValues, final Optional<String> stateVersion) {
-        this.stateValues = Collections.unmodifiableMap(stateValues == null ? Collections.emptyMap() : stateValues);
+        // Defensively copy the caller's Map so that this StateMap's contents are immutable and
+        // isolated from any subsequent mutations to the caller's reference. Extensions that retain
+        // and continue to mutate the Map they pass to setState() across invocations would otherwise
+        // race with the framework's checkpoint thread iterating the StateMap during snapshot
+        // serialization.
+        this.stateValues = (stateValues == null || stateValues.isEmpty()) ? Collections.emptyMap() : Collections.unmodifiableMap(new HashMap<>(stateValues));
         this.stateVersion = stateVersion;
     }
 

--- a/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/state/providers/local/WriteAheadLocalStateProvider.java
+++ b/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/state/providers/local/WriteAheadLocalStateProvider.java
@@ -168,6 +168,13 @@ public class WriteAheadLocalStateProvider extends AbstractStateProvider {
         return properties;
     }
 
+    // Visible for testing. Forces a synchronous checkpoint of the underlying Write-Ahead Log so
+    // that tests can deterministically observe the on-disk snapshot without waiting for the
+    // scheduled CheckpointTask to run.
+    void checkpoint() throws IOException {
+        writeAheadLog.checkpoint();
+    }
+
     @Override
     public synchronized void shutdown() {
         executor.shutdown();

--- a/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/state/providers/local/TestWriteAheadLocalStateProvider.java
+++ b/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/state/providers/local/TestWriteAheadLocalStateProvider.java
@@ -28,30 +28,91 @@ import org.apache.nifi.logging.ComponentLog;
 import org.apache.nifi.parameter.ParameterLookup;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
 import org.wali.WriteAheadRepository;
 
 import java.io.IOException;
+import java.nio.file.Path;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.Map;
-import java.util.UUID;
 import javax.net.ssl.SSLContext;
 
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
 public class TestWriteAheadLocalStateProvider extends AbstractTestStateProvider {
+    @TempDir
+    private Path temporaryDirectory;
+
     private StateProvider provider;
     private WriteAheadRepository<StateMapUpdate> wal;
 
     @BeforeEach
     public void setup() throws IOException {
-        provider = new WriteAheadLocalStateProvider();
+        provider = initializeProvider(temporaryDirectory.resolve("setup").toString());
+    }
+
+    @AfterEach
+    public void cleanup() throws IOException {
+        provider.onComponentRemoved(componentId);
+        provider.shutdown();
+
+        if (wal != null) {
+            wal.shutdown();
+        }
+    }
+
+    @Override
+    protected StateProvider getProvider() {
+        return provider;
+    }
+
+    /**
+     * Verifies that recovered state reflects the values that the caller passed to
+     * {@link StateProvider#setState(Map, String)} at the moment of that call, and is
+     * not influenced by subsequent mutations the caller performs to the same Map
+     * instance after returning. This guards against a corruption scenario where a
+     * caller (such as an extension that holds onto its own ConcurrentHashMap) keeps
+     * mutating its state map between calls to setState while the framework's
+     * checkpoint thread is concurrently serializing a snapshot. The provider must
+     * isolate its persisted state from any further mutation of the caller's map.
+     */
+    @Test
+    public void testRecoveredStateIsolatedFromPostSetStateMapMutations() throws Exception {
+        final String storageDirectory = temporaryDirectory.resolve("recovery").toString();
+        final String testComponentId = "test-mutation-isolation-component";
+
+        final WriteAheadLocalStateProvider firstProvider = initializeProvider(storageDirectory);
+        final Map<String, String> liveStateValues = new HashMap<>();
+        liveStateValues.put("key1", "value1");
+        firstProvider.setState(liveStateValues, testComponentId);
+
+        liveStateValues.put("key2", "value2");
+        liveStateValues.put("key3", "value3");
+
+        firstProvider.checkpoint();
+        firstProvider.shutdown();
+
+        final WriteAheadLocalStateProvider recoveredProvider = initializeProvider(storageDirectory);
+        try {
+            final Map<String, String> recoveredValues = recoveredProvider.getState(testComponentId).toMap();
+            assertEquals(Collections.singletonMap("key1", "value1"), recoveredValues);
+        } finally {
+            recoveredProvider.shutdown();
+        }
+    }
+
+    private WriteAheadLocalStateProvider initializeProvider(final String storageDirectory) throws IOException {
+        final WriteAheadLocalStateProvider newProvider = new WriteAheadLocalStateProvider();
         final Map<PropertyDescriptor, PropertyValue> properties = new HashMap<>();
-        properties.put(WriteAheadLocalStateProvider.PATH, new StandardPropertyValue("target/local-state-provider/" + UUID.randomUUID().toString(), null, ParameterLookup.EMPTY));
+        properties.put(WriteAheadLocalStateProvider.PATH, new StandardPropertyValue(storageDirectory, null, ParameterLookup.EMPTY));
         properties.put(WriteAheadLocalStateProvider.ALWAYS_SYNC, new StandardPropertyValue("false", null, ParameterLookup.EMPTY));
         properties.put(WriteAheadLocalStateProvider.CHECKPOINT_INTERVAL, new StandardPropertyValue("2 mins", null, ParameterLookup.EMPTY));
         properties.put(WriteAheadLocalStateProvider.NUM_PARTITIONS, new StandardPropertyValue("16", null, ParameterLookup.EMPTY));
 
-        provider.initialize(new StateProviderInitializationContext() {
+        newProvider.initialize(new StateProviderInitializationContext() {
             @Override
             public String getIdentifier() {
                 return "Unit Test Provider Initialization Context";
@@ -90,19 +151,7 @@ public class TestWriteAheadLocalStateProvider extends AbstractTestStateProvider 
                 return null;
             }
         });
-    }
 
-    @AfterEach
-    public void cleanup() throws IOException {
-        provider.onComponentRemoved(componentId);
-
-        if (wal != null) {
-            wal.shutdown();
-        }
-    }
-
-    @Override
-    protected StateProvider getProvider() {
-        return provider;
+        return newProvider;
     }
 }

